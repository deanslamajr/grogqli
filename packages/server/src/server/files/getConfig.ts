import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
