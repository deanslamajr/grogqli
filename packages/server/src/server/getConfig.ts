import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}