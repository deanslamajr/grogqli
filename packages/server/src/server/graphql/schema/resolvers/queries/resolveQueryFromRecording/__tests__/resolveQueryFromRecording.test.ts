import path from 'path';

import { resolveQueryFromRecording, ResolveQueryFromRecordingParams } from '..';
import { getConfig } from '../../../../../../getConfig';

jest.mock('../../../../../../getConfig');

describe('resolveQueryFromRecording', () => {
  beforeEach(() => {
    const mockedGetConfig = getConfig as jest.MockedFunction<typeof getConfig>;
    mockedGetConfig.mockImplementation(async () => () => {
      const relativePathToTestGrogqli = path.normalize(
        '../../../__tests__/grogqli'
      );
      return path.join(__dirname, relativePathToTestGrogqli);
    });
  });

  it('should execute the given query', async () => {
    const query =
      'query GetChores {\n  getChores {\n    chores {\n      id\n      summary\n      description\n      dueDate\n      version\n      __typename\n    }\n    hasAccountSession\n    __typename\n  }\n}\n';
    const queryConfig: ResolveQueryFromRecordingParams = {
      query,
      schemaId: 'someSchemaId',
    };
    const actual = await resolveQueryFromRecording(queryConfig);
    expect(actual).toMatchSnapshot();
  });
});
