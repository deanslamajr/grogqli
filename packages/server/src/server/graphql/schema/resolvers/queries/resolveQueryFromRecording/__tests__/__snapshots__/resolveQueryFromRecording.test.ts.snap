// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveQueryFromRecording should execute the given query 1`] = `
Object {
  "data": Object {
    "getChores": Object {
      "__typename": "GetChoresResponse",
      "chores": Array [
        Object {
          "__typename": "Chore",
          "description": "just a nice description",
          "dueDate": "10/5/1985",
          "id": "aChoreId",
          "summary": "a summary",
          "version": 1,
        },
      ],
      "hasAccountSession": true,
    },
  },
  "errors": undefined,
}
`;
